# Dataset Audit

---

## Objective

Validate dataset quality and identify potential issues before model development.

---

## Business Perspective

Poor data quality may produce inaccurate recommendations, resulting in poor customer experience and business losses.

Ensuring data quality is essential to build trustworthy recommendation systems.

---

## Data Science Perspective

Data auditing is the first technical step to understand dataset structure, identify anomalies, missing values, duplicates, and potential data leakage.

Poor quality data can negatively impact model performance.

---

## Methodology

The following checks will be performed:

1. Dataset loading
2. Dataset shape validation
3. Data type validation
4. Missing value assessment
5. Duplicate assessment
6. Descriptive statistics
7. Potential data leakage assessment

---

In [1]:
# ==================================================
# IMPORT LIBRARIES
# ==================================================

from pathlib import Path
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
# ==================================================
# DATASET LOCATION
# ==================================================

PROJECT_ROOT = Path().resolve().parent

DATA_PATH = PROJECT_ROOT / "data" / "raw" / "Online Retail.xlsx"

print(DATA_PATH)
print("File exists:", DATA_PATH.exists())

/home/egion/retail_product_recomendation/data/raw/Online Retail.xlsx
File exists: True


In [3]:
# ==================================================
# LOAD DATASET
# ==================================================

df = pd.read_excel(DATA_PATH)

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [4]:
# ==================================================
# BASIC DATASET INFORMATION
# ==================================================

print(f"Dataset Shape: {df.shape}")

display(df.head())

Dataset Shape: (541909, 8)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [6]:
# ==================================================
# MISSING VALUE CHECK
# ==================================================

missing = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percentage": round(df.isnull().mean()*100,2)
})

missing = missing.sort_values(
    by="missing_percentage",
    ascending=False
)

display(missing)

,missing_count,missing_percentage
CustomerID,135080,24.93
Description,1454,0.27
StockCode,0,0.00
InvoiceNo,0,0.00
Quantity,0,0.00
InvoiceDate,0,0.00
UnitPrice,0,0.00
Country,0,0.00


In [7]:
# ==================================================
# DUPLICATE CHECK
# ==================================================

duplicate_count = df.duplicated().sum()

print("Duplicate Rows:", duplicate_count)

Duplicate Rows: 5268


In [8]:
# ==================================================
# DATA TYPE VALIDATION
# ==================================================

dtype_df = pd.DataFrame({
    "column": df.columns,
    "dtype": df.dtypes.astype(str),
    "unique_values": [df[col].nunique() for col in df.columns]
})

display(dtype_df)

,column,dtype,unique_values
InvoiceNo,InvoiceNo,object,25900
StockCode,StockCode,object,4070
Description,Description,object,4223
Quantity,Quantity,int64,722
InvoiceDate,InvoiceDate,datetime64[us],23260
UnitPrice,UnitPrice,float64,1630
CustomerID,CustomerID,float64,4372
Country,Country,str,38


In [9]:
# ==================================================
# DESCRIPTIVE STATISTICS
# ==================================================

display(df.describe(include="all").T)

,count,unique,top,freq,mean,min,25%,50%,75%,max,std
InvoiceNo,541909.0,25900.0,573585.0,1114.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
StockCode,541909,4070,85123A,2313,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Description,540455,4223,WHITE HANGING HEART T-LIGHT HOLDER,2369,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Quantity,541909.0,NaN,NaN,NaN,9.55225,-80995.0,1.0,3.0,10.0,80995.0,218.081158
InvoiceDate,541909,NaN,NaN,NaN,2011-07-04 13:34:57.156386,2010-12-01 08:26:00,2011-03-28 11:34:00,2011-07-19 17:17:00,2011-10-19 11:27:00,2011-12-09 12:50:00,NaN
UnitPrice,541909.0,NaN,NaN,NaN,4.611114,-11062.06,1.25,2.08,4.13,38970.0,96.759853
CustomerID,406829.0,NaN,NaN,NaN,15287.69057,12346.0,13953.0,15152.0,16791.0,18287.0,1713.600303
Country,541909,38,United Kingdom,495478,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [10]:
# ==================================================
# INITIAL BUSINESS OBSERVATIONS
# ==================================================

print("Total Transactions :", df["InvoiceNo"].nunique())
print("Total Customers    :", df["CustomerID"].nunique())
print("Total Products     :", df["StockCode"].nunique())
print("Countries Covered  :", df["Country"].nunique())

Total Transactions : 25900
Total Customers    : 4372
Total Products     : 4070
Countries Covered  : 38


# Findings

1. The dataset contains 541,909 transaction records and 8 features.

2. Missing values were identified in:
   - CustomerID: 135,080 rows (24.93%)
   - Description: 1,454 rows (0.27%)

3. Duplicate rows were detected:
   - Total duplicate rows: 5,268

4. The dataset contains:
   - 25,900 unique transactions
   - 4,372 unique customers
   - 4,070 unique products
   - Transactions from 38 countries

5. Quantity and UnitPrice contain negative values, indicating potential product returns, cancellations, or correction transactions.

6. CustomerID is a critical feature for recommendation systems; therefore, missing CustomerID records require special attention.

7. The dataset appears to be highly dominated by transactions from the United Kingdom.

8. No immediate data leakage issue was identified at this stage, but transaction-based recommendation systems require careful handling during train-test splitting.

9. After removing cancellation transactions, no negative quantity records remained in the dataset.

10. This indicates that negative quantities were strongly associated with cancellation transactions.

---

# Decision

1. Records without CustomerID will be removed because customer-product interactions cannot be established without customer identifiers.

2. Duplicate rows will be investigated further during the data cleaning phase before deciding whether to remove them.

3. Transactions representing cancellations or returns will be analyzed separately to determine their suitability for recommendation modeling.

4. Negative quantities and negative prices will not be removed immediately; further business investigation is required.

5. Additional exploratory analysis will be conducted to understand customer behavior, transaction patterns, and country distribution.

6. Recommendation models will only be trained using valid customer-product interaction records.

7. Data cleaning decisions will be documented explicitly to ensure reproducibility.